src/Reserve.sol

```diff
@@ -37,44 +122,146 @@ contract Reserve is IReserve, Ownable {
+        deposited[token] += amt;
+        _transfer(token, from, _pluginAddr(plugin), amt);
+        _afterDeposition(token, plugin, amt);
         emit Deposited(msg.sender, token, from, amt);
     }
 
+    /// @notice Withdraws funds from the reserve.
+    /// The reserve will transfer `amt` tokens to the `to` address.
+    /// Only funds previously deposited can be withdrawn.
+    /// Callable only by a current user.
+    /// @param token The used token.
+    /// @param to The address to which funds are withdrawn.
+    /// @param amt The withdrawn amount.
+    function withdraw(
+        IERC20 token,
+        address to,
+        uint256 amt
+    ) public override onlyUser {
+        uint256 balance = deposited[token];
+        require(balance >= amt, "Reserve: withdrawal over balance");
+        deposited[token] = balance - amt;
+        IReservePlugin plugin = plugins[token];
+        _beforeWithdrawal(token, plugin, amt);
+        _transfer(token, _pluginAddr(plugin), to, amt);
+        emit Withdrawn(msg.sender, token, to, amt);
+    }
+
+    /// @notice Withdraws funds from the reserve or a plugin.
+    /// The reserve will transfer `amt` tokens to the `to` address.
+    /// The function doesn't update the deposited amount counter.
+    /// If used recklessly, it may cause a mismatch between the counter and the actual balance
+    /// making valid future calls to `withdraw` or `setPlugin` fail due to lack of funds.
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param plugin The plugin to withdraw from.
+    /// It doesn't need to be registered as a plugin for `token`.
+    /// Pass `NO_PLUGIN` to withdraw directly from the reserve balance.
+    /// @param to The address to which funds are withdrawn.
+    /// @param amt The withdrawn amount.
     function forceWithdraw(
         IERC20 token,
+        IReservePlugin plugin,
         address to,
         uint256 amt
     ) public onlyOwner {
-        require(token.transfer(to, amt), "Reserve: transfer failed");
-        emit ForceWithdrawn(msg.sender, token, to, amt);
+        _beforeWithdrawal(token, plugin, amt);
+        _transfer(token, _pluginAddr(plugin), to, amt);
+        emit ForceWithdrawn(msg.sender, token, plugin, to, amt);
+    }
+
+    /// @notice Sets the deposited amount counter for a token without transferring any funds.
+    /// If used recklessly, it may cause a mismatch between the counter and the actual balance
+    /// making valid future calls to `withdraw` or `setPlugin` fail due to lack of funds.
+    /// It may also make the counter lower than what users expect it to be again making
+    /// valid future calls to `withdraw` fail.
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param amt The new deposited amount counter value.
+    function setDeposited(IERC20 token, uint256 amt) public onlyOwner {
+        deposited[token] = amt;
```

Comment:
Do we really need this function? If we implement some surplus harvest pattern?

Reply:
Yes, we still do, there needs to be an emergency tool so if something goes awfully wrong, we can manually fix the state.

Reply:
Yes, i can see the concern. Lets keep it, just to have a emergency mechanism.  The governance anyway has access to the funds as well.